google/cloud/internal/async_rest_polling_loop.h

```diff
@@ -0,0 +1,137 @@
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_ASYNC_REST_POLLING_LOOP_H
+
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/future.h"
+#include "google/cloud/internal/rest_context.h"
+#include "google/cloud/polling_policy.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <google/longrunning/operations.pb.h>
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+using AsyncRestPollLongRunningOperation =
+    std::function<future<StatusOr<google::longrunning::Operation>>(
+        google::cloud::CompletionQueue&, RestContext&,
+        google::longrunning::GetOperationRequest const&)>;
+
+using AsyncRestCancelLongRunningOperation = std::function<future<Status>(
+    google::cloud::CompletionQueue&, RestContext&,
+    google::longrunning::CancelOperationRequest const&)>;
+
+/**
+ * Runs an asynchronous polling loop for a long-running operation.
+ *
+ * Long-running operations [aip/151] are used for API methods that take a
+ * significant amount of time to complete (think minutes, maybe an hour). The
+ * API returns a "promise" object, represented by the
+ * `google::longrunning::Operation` proto, and the application (or client
+ * library) should periodically poll this object until it is "satisfied".
+ *
+ * This function runs an asynchronous loop to poll the long-running operation.
+ * It periodically invokes the @p poll function to query the status of the
+ * operation. If the operation has completed (even if completed with an error)
+ * the loop stops. If the operation has not completed, or polling fails (say
+ * because the network has a problem), the function sets an asynchronous timer
+ * as configured by the polling policy and tries again later. The polling policy
+ * can stop the loop too.
+ *
+ * The function returns a `future<>` that is satisfied when the loop stops. In
+ * short, the returned future is satisfied under any of these conditions (a) the
+ * polling policy is exhausted before it is known if the operation completed
+ * successfully, or (b) the operation completes, and this is known because a
+ * `GetOperation()` request returns the operation result.
+ *
+ * The promise can complete with an error, which is represented by a
+ * `google::cloud::Status` object, or with success and some `ReturnType` value.
+ * The application may also configure the "polling policy", which may stop the
+ * polling even though the operation has not completed.
+ *
+ * Usually the initial request to "start" the long-running operation also
+ * requires a retry loop, which is handled by this function too.
+ *
+ * Typically, library developers would use the function via
+ * `AsyncLongRunningOperation`, but as a stand-alone function it can be used in
+ * this context. First assume there is a `*Stub` class as follows:
+ *
+ * @code
+ * class BarStub {
```

Comment:
optional nit: s/BarStub/BarRestStub/

And maybe the same thing below with `BarConnection`?